src/topology/path_connected.lean

```diff
@@ -0,0 +1,518 @@
+
+* `joined (x y : X)` means there is a continuous `γ : I → X`, such that `γ 0 = x` and `γ 1 = y`.
+* `path_component (x : X)` is the set of points joined to `x`.
+* `path_connected_space X` is a predicate class asserting that `X` is non-empty and every two
+  points of `X` are joined.
+
+Then there are corresponding relative notions for `F : set X`.
+
+* `joined_in F (x y : X)` means there is a continuous `γ : I → X` with values in `F`,
+  such that `γ 0 = x` and `γ 1 = y`.
+* `path_component_in F (x : X)` is the set of points joined to `x` in `F`.
+* `is_path_connected F` asserts that `F` is non-empty and every two
+  points of `F` are joined in `F`.
+* `loc_path_connected_space X` is a predicate class asserting that `X` is locally path-connected:
+  each point has a basis of path-connected neighborhoods (we do *not* ask these to be open).
+
+## Main theorems
+
+* `joined` and `joined_in F` are transitive relations.
+
+One can link the absolute and relative version in two direction, using `(univ : set X)` or the
+subtype `↥F`.
+
+* `path_connected_space_iff_univ : path_connected_space X ↔ is_path_connected (univ : set X)`
+* `is_path_connected_iff_path_connected_space : is_path_connected F ↔ path_connected_space ↥F`
+
+For locally path connected spaces, we have
+* `path_connected_space_iff_connected_space : path_connected_space X ↔ connected_space X`
+* `is_connected_iff_is_path_connected (U_op : is_open U) : is_path_connected  U ↔ is_connected U`
+
+## Implementation notes
+
+By default, all paths have `I` as their source and `X` as their target, but there is an
+operation `I_extend` that will extend any continuous map `γ :I → X` into a continuous map
+`I_extend γ : ℝ → X` that is constant before `0` and after `1`.
+
+This is used to define `joined.extend` that turns `h : joined x y` into a continuous map
+`h.extend : ℝ → X` whose restriction to `I` joins `x` and `y`.
+
+Similarly, one can turn `h : joined_in F x y` into a continuous map
+`h.extend : ℝ → X`, and `h.map : I → F` taking values in the subtype `F`,
+and also `h.extend_map : ℝ → F`.
+
+-/
+
+noncomputable theory
+open_locale classical topological_space filter
+open filter set
+
+variables {X : Type*} [topological_space X] {x y z : X} {ι : Type*}
+
+local notation `I` := Icc (0 : ℝ) 1
+
+lemma Icc_zero_one_symm {t : ℝ} : t ∈ I ↔ 1 - t ∈ I :=
+begin
+  rw [mem_Icc, mem_Icc],
+  split ; intro ; split ; linarith
+end
+
+instance I_has_zero : has_zero I := ⟨⟨0, by split ; norm_num⟩⟩
+
+@[simp, norm_cast] lemma coe_I_zero : ((0 : I) : ℝ) = 0 := rfl
+
+instance I_has_one : has_one I := ⟨⟨1, by split ; norm_num⟩⟩
+
+@[simp, norm_cast] lemma coe_I_one : ((1 : I) : ℝ) = 1 := rfl
+
+/-- Unit interval central symmetry. -/
+def I_symm : I → I := λ t, ⟨1 - t.val, Icc_zero_one_symm.mp t.property⟩
+
+local notation `σ` := I_symm
+
+@[simp] lemma I_symm_zero : σ 0 = 1 :=
+subtype.ext $ by simp [I_symm]
+
+@[simp] lemma I_symm_one : σ 1 = 0 :=
+subtype.ext $ by simp [I_symm]
```

Comment:
Lemmas up to this one probably should go to some file in `data/set/intervals`. They are useful, e.g., for `analysis/convex` as well.

Reply:
I don't see where I could use those lemmas in `analysis/convex`. Note that `analysis.convex.basic` defines the local notation `I` but uses it only three times. 

I could put everything not using the topology on `I` in `data/set/intervals/unit_interval.lean` but then where do I put the continuity lemmas? It would be weird to have them super far apart. It will obviously need `topology.instances.real` which is the only dependency of the current file.

Overall I'm not convinced the split makes much sense.